src/IIIFPresentation/API/Features/Manifest/CanvasPaintingResolver.cs

```diff
@@ -0,0 +1,164 @@
+        int customerId, IIIFManifest presentationManifest, CancellationToken cancellationToken)
+    {
+        var canvasPaintings = manifestItemsParser.ParseItemsToCanvasPainting(presentationManifest).ToList();
+        var insertCanvasPaintingsError = await HandleInserts(canvasPaintings, customerId, cancellationToken);
+        if (insertCanvasPaintingsError != null) return (insertCanvasPaintingsError, null);
+
+        return (null, canvasPaintings);
+    }
+
+    /// <summary>
+    /// Reconcile incoming manifest with any CanvasPainting objects already stored in DB. Resulting DB records should
+    /// reflect incoming manifest.
+    /// </summary>
+    public async Task<PresUpdateResult?> UpdateCanvasPaintings(int customerId, IIIFManifest presentationManifest,
+        DbManifest existingManifest, CancellationToken cancellationToken)
+    {
+        var incomingCanvasPaintings =
+            manifestItemsParser.ParseItemsToCanvasPainting(presentationManifest).ToList();
+
+        existingManifest.CanvasPaintings ??= [];
+
+        // Iterate through all incoming - this is what we want to preserve in DB
+        var processedCanvasPaintingIds = new List<int>(incomingCanvasPaintings.Count);
+        var toInsert = new List<CanvasPainting>();
+        foreach (var incoming in incomingCanvasPaintings)
+        {
+            CanvasPainting? matching = null;
+            var candidates = existingManifest.CanvasPaintings
+                .Where(cp => cp.CanvasOriginalId == incoming.CanvasOriginalId).ToList();
+            if (candidates.Count == 1)
+            {
+                // Single item matching - check if we've processed it already. If so this is due to choice
+                var potential = candidates.Single();
+                if (!processedCanvasPaintingIds.Contains(potential.CanvasPaintingId))
+                {
+                    logger.LogTrace("Found existing canvas painting for {CanvasOriginalId}", incoming.CanvasOriginalId);
+                    matching = potential;
+                }
+            }
+            else if (candidates.Count > 1)
+            {
+                // If there are multiple matching items then Canvas is a choice
+                logger.LogTrace("Found multiple canvas paintings for {CanvasOriginalId}", incoming.CanvasOriginalId);
+                matching = candidates.SingleOrDefault(c => c.ChoiceOrder == incoming.ChoiceOrder);
+            }
+
+            if (matching == null)
+            {
+                if (incoming.ChoiceOrder.HasValue)
+                {
+                    // This is a choice. If there are other, existing items for the same canvas then seed canvas_id
+                    incoming.Id = candidates.SingleOrDefault()?.Id;
+                }
+
+                // Store it in a list for processing later (e.g. for bulk generation of UniqueIds)
+                toInsert.Add(incoming);

```

Comment:
What would be a valuable message to log that actually tells something useful? `canvasOriginalId`?  `canvas_order`, `choice_order`

Reply:
potentially the `CanvasPaintingId`? - I was thinking more in lines that every other path in this method is logged, so it might be worth it to get this too - I'm pretty happy for the answer to that question to be "not useful" though

Reply:
I think it would be useful, I just wasn't sure what would be a succinct enough message. `CanvasPaintingId` isn't known until record is saved in DB so wont' be known here.

Maybe the fact that a canvas will be added at index X is enough (for now anyway). `logger.LogTrace("Adding canvas {CanvasIndex}, choice {ChoiceIndex}", matching.CanvasIndex, matching.ChoiceIndex);`